### DISCUSSION TITLE
Best way to modularly display a grid

### DISCUSSION THREAD

**Discussion by Tmpecho** [slint-ui/slint#5422] [Q&A]
For my app, I have rooms with some information in a struct about them that are on different floors. These rooms can change for different building, so I want to find a way to based on for example a 2d list, display the rooms in a grid view (or for example a vertical layout with a horizontal layout in each vertical layout) where a row is rooms in a floor. 
I've found that it's difficult to do this as for example the `GridBox` don't allow for loops. What i'd like to do is writing something like this (extremely roughly):
```slint
Grid {  // imaginary component
    in property <[ [room-struct] ]> floors-data;
    height: 100%;
    width: 100%;
    for floor in floors-data {
        for room in floor {
            Rectangle {
                Room { room-data: room }
            }
        }
    }
}
```
For context, the room names are now like `["101", "102", "201", "202"]`.

Another way i've tried to do is like this:
```slint
...
VerticalBox {
    for floor in floors-data: HorizontalBox {
        for room-name in floor: Room {
            room-name: room-name;
        }
    }
}
```
However, then I need to make the floors-data in the correct way in rust, which has proven hard to me with my limited Slint knowledge. I was able to use rust to make a list of `StandardListViewItem` items to use in a `StandardTableView`, but this was easier to do, as I was able to do it.

The only help i've tried to now is chatGPT, which has limited Slint knowledge, so it has unremarkably made tons of mistakes, and often hallucinates, as can be expected.

So my question would be if there is an easy way to do this; take a vector of a struct (or string for simplicity's sake) and make that into a grid view.

I can provide my attempt with more specific code details if anyone would be so kind to aid me in this issue.

Thanks!

**Comment by ogoffart** [comment]
our GridLayout does not yet support iteration, so one way to do it is to implement grid yourself by setting the geometry manually.
Examples:
https://github.com/slint-ui/slint/blob/e232648d019216904dbe67498632a4dd52705eaa/examples/printerdemo/ui/home_page.slint#L53-L69

https://github.com/slint-ui/slint/blob/e232648d019216904dbe67498632a4dd52705eaa/examples/memory/memory.slint#L106-L117

**Comment by LouisBouch** [comment]
It's been a while, but I'll leave this here anyway in case someone else stumbles upon this.

I got this solution by adapting the Rust code from the [tutorial](https://docs.slint.dev/latest/docs/slint/tutorial/quickstart/) and with Claude as a helping hand.

I managed to make a grid by placing a bunch of `HorizontalLayout` within a `VerticalLayout`.

## Slint code
Multiple `HorizontalLayout` inside a `VerticalLayout`:
```slint
in property <[[TileData]]> memory_tiles: [[...]]; // Put your data in ...
VerticalLayout {
    for tiles_in_row[r] in memory_tiles: HorizontalLayout {
        for tile[i] in tiles_in_row: MemoryTile {
            <YOUR CODE>
        }
    }
}
```
## Rust code
In Rust, you can fetch the 2D table data, transform it, and send it back to Slint:
```Rust
slint::include_modules!();
use slint::Model;

fn main() {
    let main_window = MainWindow::new().unwrap();

    // Fetch the tiles from the model
    let mut tiles: Vec<Vec<TileData>> = main_window
        .get_memory_tiles()
        .iter()
        .map(|inner_model| inner_model.iter().collect())
        .collect();
    
    let mut grid: Vec<Vec<TileData>>= tiles.clone();
    //
    // Modify your grid in any way you see fit.
    //

    // Send it back to Slint.
    let grid_model = slint::ModelRc::new(slint::VecModel::from(
        grid.into_iter()
            .map(|inner_vec| slint::ModelRc::new(slint::VecModel::from(inner_vec)))
            .collect::<Vec<_>>(),
    ));
    main_window.set_memory_tiles(grid_model.clone().into());

    main_window.run().unwrap();
}

```

## Comparison to tutorial code
There are 2 mains differences from the code in the tutorial and the one provided above.

### Fetching data
To fetch the data from the slint component, the [tutorial](https://docs.slint.dev/latest/docs/slint/tutorial/creating_the_tiles/) uses 
```Rust
let mut tiles: Vec<TileData> = main_window.get_memory_tiles().iter().collect();
```
 Essentially, `main_window.get_memory_tiles()` has type `ModelRc<TileData>` and we can iterate over its `TileData` element in order to build a Rust Vector. The `collect()` call automatically knows to convert the values into a `Vec<TileData>` given the type annotation.
 
 In our case, to accommodate the 2D nature of `main_window.get_memory_tiles()`, which now has type `ModelRc<ModelRc<TileData>>`, we do instead:
``` Rust
 let mut tiles: Vec<Vec<TileData>> = main_window
        .get_memory_tiles()
        .iter()
        .map(|inner_model| inner_model.iter().collect())
        .collect();
```
Notice that the only difference between the two snippets of code is the `map`. This mapping converts the inner tables into valid rust Vectors. This is required given that we are working with 2D vectors and not a 1D vector.

### Updating slint 2D table
At then end, to update the slint table, we execute:
```Rust
    let grid_model = slint::ModelRc::new(slint::VecModel::from(
        grid.into_iter()
            .map(|inner_vec| slint::ModelRc::new(slint::VecModel::from(inner_vec)))
            .collect::<Vec<_>>(),
    ));
    main_window.set_memory_tiles(grid_model.clone().into());

```
Which is different from what is used in the [tutorial](https://docs.slint.dev/latest/docs/slint/tutorial/creating_the_tiles/):
```Rust
    let tiles_model = std::rc::Rc::new(slint::VecModel::from(tiles));
    main_window.set_memory_tiles(tiles_model.clone().into());
```

There are 2 differences here.
1. We must add a `map` to convert the inner Vectors.
2. We use `slint::ModelRc::new(slint::VecModel::from(...)` instead of `std::rc::Rc::new(slint::VecModel::from(...)`.

**1.** The first difference is very similar to what we did in the first section when fetching the table from slint to accommodate the 2D nature of the data. Instead of converting only the outer Vector, we also iterate over every single inner Vector and execute the conversion.

**2.** As for why we used `slint::ModelRc::new(slint::VecModel::from(...)` instead of `std::rc::Rc::new(slint::VecModel::from(...)` to make our smart pointers, I'm not too sure. My best bet is that slint implicitely converts Rust's `Rc<T>` smart pointer into its own smart pointer `ModelRc<T>`, but when we have nested tables, you must explicitly use Slint's smart pointers, which are created using `slint::ModelRc::new(slint::VecModel::from(...))`. Note that in the 1D scenario, you can use both method interchangeably. 

I hope this helps whoever looks at this in the future.
